enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,468 @@
+perform an egress IP re-assignment every time a route goes down).
+
+Given the above, it is however important that the network plugin is conveyed the
+cloud's information concerning the nodes' subnet range. This is especially
+evident on GCP, as without it: the network plugin won't be able to perform any
+assignments at all, given the nodes' default `/32` subnet.
+
+##### OpenShift vs. upstream project
+
+There are two approaches that can be taken for hosting this component. Normally,
+this component would be an OpenShift dedicated component, however since
+ovn-org/ovn-kubernetes does implement the egress IP feature supported on
+bare-metal clusters: it has been discussed whether supporting the cloud egress
+IP feature would make sense for that organization as well.
+
+**upstream**
+
+An "upstream model" would entail getting the sign-off from ovn-org that such a
+feature is useful. Pending a green light, an OpenShift dedicated downstream
+repository would need to be created following that.
+
+The cost of such an approach is more time spent in setting up CI/CD for both
+repositories.
+
+##### CRD vs. a annotated approach
+
+Multiple pieces of information needs to be communicated between the network
+plugins and this cloud component:
+
+- The cloud's node subnet information.
+- The egress IP assignment to be performed by the cloud component
+
+and possibly:
+
+- The status of the egress IP assignment performed by the cloud component
+
+Two models have been imagined for this: a CRD based approach, or a node
+annotated approach.
+
+**CRD**
+
+The cloud component could use a dedicated CRD that the network plugin watches.
+The CRD could store the nodes' subnet, as known by the cloud. The network plugin
+could also use this CRD to issue requests for cloud egress IP assignments, and
+the cloud component could update a status field indicating a successful cloud
+assignment of the egress IP.
+
+Doing this would however introduce explicit dependencies between the network
+plugin and this cloud component. If the component ends up being a dedicated
+OpenShift one: the CRD would end up creating a awkward dependency tree between
+ovn-org and OpenShift, since this feature would be implemented upstream in
+ovn-org/ovn-kubernetes. If an upstream model is chosen the reverse happens: we
+will have ovn-org dedicated CRD being watched for by openshift-sdn.
+
+**annotated approach**
+
+The second approach imagined is annotating the Kubernetes node object. The cloud
+component would query the cloud API for the nodes' subnet range and annotate the
+node objects according to a defined convention which the network plugins can
+read. That information would then be taken into account during the plugin's
+assignment procedure.
+
+Vice versa, the network plugin could issue a "cloud egress IP assignment
+request" by means of annotating the node object. The cloud component could use
+this annotation as input to start executing an IP assignment in the cloud.
```

Comment:
> It doesn't make sense adding an annotation on all the nodes in the cluster specially when we don't even know if we will use it.

Yes, we should be:

- Keeping the "cloud subnet annotation" (`cloud.api/cloud-primary-ifaddr`) split from the "OS subnet annotation" (`k8s.ovn.org/node-primary-ifaddr`) as to ensure a clean design and backward compatibility for the network plugin with versions of the code that never had this cloud component. 
- Annotating all nodes with the "cloud subnet annotation" at startup since we don't know which nodes the user will want to use for egress IP, and this shouldn't cost a lot to do for the cloud component. 

I can add this to the enhancement proposal